subsys: net: lib: add wifi_credentials library

maxd-nordic opened the pull request:
Upstream NCS's library for storing Wi-Fi credentials.
This library allows storage of Wi-Fi credentials using different backends.
Either the Zephyr settings subsystem or the PSA secure backend can be used.
For testing purposes, credentials can be defined statically.

maxd-nordic committed:
net: lib: add wifi_credentials library

Upstream NCS's library for storing Wi-Fi credentials.
This library allows storage of Wi-Fi credentials
using different backends.
Either the Zephyr settings subsystem
or the PSA secure backend can be used.
For testing purposes, credentials can be defined statically.

Signed-off-by: Ravi Dondaputi <ravi.dondaputi@nordicsemi.no>
Signed-off-by: Kapil Bhatt <kapil.bhatt@nordicsemi.no>
Signed-off-by: Gregers Gram Rygg <gregers.gram.rygg@nordicsemi.no>
Signed-off-by: Kaja Koren <kaja.koren@nordicsemi.no>
Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>
Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

maxd-nordic commented:
@ioannisg @krish2718

rlubos commented on subsys/net/lib/wifi_credentials/wifi_credentials.c:
```diff
@@ -0,0 +1,398 @@
+
+LOG_MODULE_REGISTER(wifi_credentials, CONFIG_WIFI_CREDENTIALS_LOG_LEVEL);
+
+K_MUTEX_DEFINE(wifi_credentials_mutex);
```
nit: static

rlubos reviewed (commented):
Please check the compliance results, there are a few failures about unrecognized Kconfig symbols and variable naming. Also, no `subsys: ` in the commit title prefix please.

The commit message could elaborate a bit about the subsystem purpose, not only mention that it was contributed from NCS.

But overall, nice! Any plans to port tests (I see it might be a challenge given they're unity-based)? Also, the wifi_mgmt extension from NCS, utilizing this library to connect, would be a nice addition to this.

maxd-nordic commented:
@rlubos Thanks! I'll port the tests and the autoconnect feature as well. Do you know what the Coding Guideline issue is about?

rlubos commented:
> Do you know what the Coding Guideline issue is about?

The variable name is required to be different from the type (struct) name. We typically do something like `const struct shell *sh`.

maxd-nordic committed:
net: lib: wifi_credentials: add connect-stored feature

This patch adds a feature to directly connect to stored Wi-Fi
credentials without having to compose the NET_MGMT commands yourself.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

jukkar commented on include/zephyr/net/wifi_credentials.h:
```diff
@@ -0,0 +1,218 @@
+ *
+ */
+struct wifi_credentials_header {
+	enum wifi_security_type type;
+	char ssid[WIFI_SSID_MAX_LEN];
+	size_t ssid_len;
+	uint8_t bssid[WIFI_MAC_ADDR_LEN];
+	uint32_t flags;
+	uint8_t channel;
+	uint32_t timeout;
```
This looks like it will leave a hole in memory. If you organize the fields differently we can save couple of bytes.
Suggesting something like this
```
	enum wifi_security_type type;
	char ssid[WIFI_SSID_MAX_LEN];
	size_t ssid_len;
	uint32_t flags;
	uint32_t timeout;
	uint8_t bssid[WIFI_MAC_ADDR_LEN];
	uint8_t channel;
```
Also the fields need to be documented, otherwise the generated documentation will look like crap.
Same comment about field documentation to other structs below.

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials.c:
```diff
@@ -0,0 +1,398 @@
+	/* load entry */
+	ret = wifi_credentials_get_by_ssid_personal_struct(ssid, ssid_len,
+							   (struct wifi_credentials_personal *)buf);
+
```
empty line can be removed here

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials_backend_psa.c:
```diff
@@ -0,0 +1,97 @@
+{
+	size_t length_read = 0;
+	size_t key_id = idx + WIFI_CREDENTIALS_BACKEND_PSA_KEY_ID_USER_MIN;
+
+	psa_status_t ret = psa_export_key(key_id, buf, buf_len, &length_read);
+
+	if (ret != PSA_SUCCESS) {
```
I suggest formatting this differently, it looks a bit funny now. So something like
```
size_t key_id = ....;
psa_status_t ret;

ret = psa_export_key(key_id, buf, buf_len, &length_read);
if (ret != ....
```

maxd-nordic committed:
tests: net: lib: wifi_credentials: add tests

This patch adds tests for the wifi_credentials library.

Both the main functions and the backend implementations are tested.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

rlubos reviewed (commented):
Please use `net: lib: ` prefix in the second commit  and `tests: net: lib` in the test commit.

But otherwise, nice!

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials_connect_stored.c:
```diff
@@ -0,0 +1,238 @@
+	return ret;
+};
+
+NET_MGMT_REGISTER_REQUEST_HANDLER(NET_REQUEST_WIFI_CONNECT_STORED, wifi_ext_command);
```
Also do we need a separate shell for these, shouldn't the shell functionality in this PR be part of wifi shell?

jukkar replied:
@rlubos yeah, I think you are right, the wifi mgmt commands should go there.

jukkar replied:
What about my shell question? The implementation could stay in this .c file, but the credentials could be found under the wifi shell.

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials_connect_stored.c:
```diff
@@ -0,0 +1,238 @@
+#include <zephyr/net/wifi_mgmt.h>
+#include <stdio.h>
+
+/* Ensure 'strnlen' is available even with -std=c99. */
+size_t strnlen(const char *buf, size_t bufsz);
```
No need to have this here, just set
```
#undef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE 200809L /* For strnlen() */
```
before the include statements.

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials_connect_stored.c:
```diff
@@ -0,0 +1,238 @@
+		ret = -ENOMEM;
+		goto err_out;
+	}
+	memset(ssid, 0, creds->header.ssid_len + 1);
```
Please add empty lines after `if {}` blocks here and everywhere in this PR.

jukkar commented on subsys/net/lib/wifi_credentials/wifi_credentials_shell.c:
```diff
@@ -290,3 +287,23 @@ SHELL_SUBCMD_ADD((wifi_cred), delete, NULL, "Delete network from storage.", cmd_
+	return 0;
+}
+
+SHELL_SUBCMD_ADD((wifi_cred), auto_connect, NULL, "Connect to any stored network.",
```
If we change `(wifi_cred)` to `(wifi)` then the commands should be found under the wifi shell instead of creating a new shell.

maxd-nordic commented:
@jukkar @rlubos I did the requested changes. If you are fine with how it looks now, I'll clean up and rebase.

rlubos commented:
@maxd-nordic Thanks for the changes, for me looks good!

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -9,6 +9,9 @@
  * @brief WiFi shell module
```
I would prefer that the keep the source code of the shell commands in wifi_credentials directory and the file. The wifi_shell.c is quite large already so don't want to clutter it with extra stuff. Sorry that I did not manage to explain myself clearly.
So I did similar thing with the bridge shell, and moved the shell **functionality** under the net shell but left the code in where it is now. See this commit https://github.com/zephyrproject-rtos/zephyr/pull/77235/commits/54cd26e1a162cb2719a9d5c5df6fa4e5afa457cd for an example.
Similar thing could be done with wifi credentials.

maxd-nordic replied:
@jukkar @rlubos I tried to do that. It requires a bigger change. What do you think?

jukkar replied:
I don't think the change is that big. The `wifi_commands` in `wifi_shell.c` needs to be split into individual pieces, meaning for each subcommand a `SHELL_SUBCMD_ADD()` needs to created.
I did an experiment and got following diff, it converted the "wifi ap" command, but I did not touch the others in wifi_commands.
```
diff --git a/subsys/net/l2/wifi/wifi_shell.c b/subsys/net/l2/wifi/wifi_shell.c
index 64bee4035ae..9d3df0b7161 100644
--- a/subsys/net/l2/wifi/wifi_shell.c
+++ b/subsys/net/l2/wifi/wifi_shell.c
@@ -2692,6 +2692,17 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_cmd_ap,
 	SHELL_SUBCMD_SET_END
 );
 
+static int cmd_wifi_ap(const struct shell *sh, size_t argc, char *argv[])
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	return 0;
+}
+
+SHELL_SUBCMD_ADD((wifi), ap, &wifi_cmd_ap,
+		 "Access Point mode commands.", cmd_wifi_ap, 0, 0);
+
 SHELL_STATIC_SUBCMD_SET_CREATE(wifi_twt_ops,
 	SHELL_CMD_ARG(quick_setup, NULL, " Start a TWT flow with defaults:\n"
 		"<twt_wake_interval: 1-262144us> <twt_interval: 1us-2^31us>.\n",
@@ -2798,7 +2809,6 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 	SHELL_CMD_ARG(version, NULL, "Print Wi-Fi Driver and Firmware versions\n",
 		  cmd_wifi_version,
 		  1, 0),
-	SHELL_CMD(ap, &wifi_cmd_ap, "Access Point mode commands.\n", NULL),
 	SHELL_CMD_ARG(connect, NULL,
 		  "Connect to a Wi-Fi AP\n"
 		  "<-s --ssid \"<SSID>\">: SSID.\n"
@@ -2936,7 +2946,9 @@ SHELL_STATIC_SUBCMD_SET_CREATE(wifi_commands,
 	SHELL_SUBCMD_SET_END
 );
 
-SHELL_CMD_REGISTER(wifi, &wifi_commands, "Wi-Fi commands", NULL);
+SHELL_SUBCMD_SET_CREATE(wifi_cmds, (wifi));
+
+SHELL_CMD_REGISTER(wifi, &wifi_cmds, "Wi-Fi commands", NULL);
 
 static int wifi_shell_init(void)
 {
diff --git a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
index ef025982326..72c78281299 100644
--- a/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
+++ b/subsys/net/lib/wifi_credentials/wifi_credentials_shell.c
@@ -289,18 +289,17 @@ static int cmd_list_networks(const struct shell *sh, size_t argc, char *argv[])
 	return 0;
 }
 
-SHELL_SUBCMD_SET_CREATE(sub_wifi_cred, (wifi_cred));
-SHELL_SUBCMD_ADD((wifi_cred), add, NULL, "Add network to storage.", cmd_add_network, 0, 0);
-SHELL_SUBCMD_ADD((wifi_cred), delete, NULL, "Delete network from storage.", cmd_delete_network, 0,
-		 0);
-SHELL_SUBCMD_ADD((wifi_cred), list, NULL, "List stored networks.", cmd_list_networks, 0, 0);
-
-SHELL_CMD_REGISTER(wifi_cred, &sub_wifi_cred, "Wi-Fi Credentials commands", NULL);
-
-#if CONFIG_WIFI_CREDENTIALS_CONNECT_STORED
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_wifi_cred,
+       SHELL_CMD(add, NULL, "Add network to storage.", cmd_add_network),
+       SHELL_CMD(delete, NULL, "Delete network from storage.", cmd_delete_network),
+       SHELL_CMD(list, NULL, "List stored networks.", cmd_list_networks),
+       SHELL_SUBCMD_SET_END
+);
 
 static int cmd_auto_connect(const struct shell *sh, size_t argc, char *argv[])
 {
+#if CONFIG_WIFI_CREDENTIALS_CONNECT_STORED
+
 	struct net_if *iface = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));
 	int rc = net_mgmt(NET_REQUEST_WIFI_CONNECT_STORED, iface, NULL, 0);
 
@@ -309,11 +308,10 @@ static int cmd_auto_connect(const struct shell *sh, size_t argc, char *argv[])
 			    "An error occurred when trying to auto-connect to a network. err: %d",
 			    rc);
 	}
-
-	return 0;
-}
-
-SHELL_SUBCMD_ADD((wifi_cred), auto_connect, NULL, "Connect to any stored network.",
-		 cmd_auto_connect, 0, 0);
-
 #endif /* CONFIG_WIFI_CREDENTIALS_CONNECT_STORED */
+
+	return 0;
+}
+
+SHELL_SUBCMD_ADD((wifi), cred, &sub_wifi_cred, "Connect to any stored network.",
+		 cmd_auto_connect, 0, 0);
```
I am not sure how the `auto_connect` is suppose to work, but just did a quick hack to get things working.

I am pushing this change as I would like the wifi_shell.c to be split into smaller pieces like what we did for net-shell. So all this work is preparation for that. It is causing problems if we have "wifi_cred" shell and then later move it to wifi shell and need to change the documentation etc.

maxd-nordic commented:
@rlubos @jukkar I cleaned up the history. Would you take another look?

rlubos commented:
Compliance nits:
```
 OPEN_BRACE: that open brace { should be on the previous line
File:subsys/net/l2/wifi/wifi_mgmt.c
Line:1169
 LONG_LINE: line length of 102 exceeds 100 columns
File:subsys/net/l2/wifi/wifi_mgmt.c
Line:1212
 OPEN_BRACE: that open brace { should be on the previous line
File:tests/net/lib/wifi_credentials_backend_psa/src/main.c
Line:55
 OPEN_BRACE: that open brace { should be on the previous line
File:tests/net/lib/wifi_credentials_backend_psa/src/main.c
Line:68
 OPEN_BRACE: that open brace { should be on the previous line
File:tests/net/lib/wifi_credentials_backend_settings/src/main.c
Line:52
 OPEN_BRACE: that open brace { should be on the previous line
File:tests/net/lib/wifi_credentials_backend_settings/src/main.c
Line:65
```

jukkar commented:
@maxd-nordic The wifi-shell changes are now in upstream main, could you rebase the PR and fix the compliance issue and then re-submit as this is ready to go in.
